daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Event.scala

```diff
@@ -112,7 +112,8 @@ object Event {
     }
 
     def mapContractIdAndValue[Cid2, Val2](f: Cid => Cid2, g: Val => Val2): Events[Nid, Cid2, Val2] =
-      copy(events = events.mapValues(_.mapContractId(f, g)))
+      // do NOT use `Map#mapValues`! it applies the function lazily on lookup. see #1861
+      copy(events = events.transform { case (_, value) => value.mapContractId(f, g) })
```

Comment:
```suggestion
      copy(events = events.transform { (_, value) => value.mapContractId(f, g) })
```

and likewise elsewhere. (Scala does not optimize this, it does exactly the silliest thing you can imagine.)

Reply:
done.